app/app.go

```diff
@@ -456,209 +475,101 @@ func New(
+	return res, nil
 }
 
-// Info implements the ABCI interface. This method is a wrapper around baseapp's
-// Info command so that it can take the app version and setup the multicommit
-// store.
-//
-// Side-effect: calls baseapp.Init()
-func (app *App) Info(req abci.RequestInfo) abci.ResponseInfo {
-	if height := app.LastBlockHeight(); height > 0 {
-		ctx, err := app.CreateQueryContext(height, false)
-		if err != nil {
-			panic(err)
-		}
-		appVersion := app.GetAppVersionFromParamStore(ctx)
-		if appVersion > 0 {
-			app.SetAppVersion(ctx, appVersion)
-		} else {
-			app.SetAppVersion(ctx, v1)
-		}
+// InitChainer is middleware that gets invoked part-way through the baseapp's InitChain invocation.
+func (app *App) InitChainer(ctx sdk.Context, req *abci.RequestInitChain) (*abci.ResponseInitChain, error) {
+	var genesisState GenesisState
+	if err := tmjson.Unmarshal(req.AppStateBytes, &genesisState); err != nil {
+		return nil, err
 	}
-
-	resp := app.BaseApp.Info(req)
-	// mount the stores for the provided app version
-	if resp.AppVersion > 0 && !app.IsSealed() {
-		app.mountKeysAndInit(resp.AppVersion)
+	versionMap := app.ModuleManager.GetVersionMap()
+	if err := app.UpgradeKeeper.SetModuleVersionMap(ctx, versionMap); err != nil {
+		return nil, err
 	}
 
-	resp.Timeouts.TimeoutCommit = app.getTimeoutCommit(resp.AppVersion)
-	resp.Timeouts.TimeoutPropose = appconsts.GetTimeoutPropose(resp.AppVersion)
-
-	return resp
+	return app.ModuleManager.InitGenesis(ctx, app.AppCodec(), genesisState)
 }
 
-// InitChain implements the ABCI interface. This method is a wrapper around
-// baseapp's InitChain so that we can take the app version and setup the multicommit
-// store.
-//
-// Side-effect: calls baseapp.Init()
-func (app *App) InitChain(req abci.RequestInitChain) (res abci.ResponseInitChain) {
-	req = setDefaultAppVersion(req)
-	appVersion := req.ConsensusParams.Version.AppVersion
-	// mount the stores for the provided app version if it has not already been mounted
-	if app.AppVersion() == 0 && !app.IsSealed() {
-		app.mountKeysAndInit(appVersion)
-	}
-
-	res = app.BaseApp.InitChain(req)
-
-	ctx := app.NewContext(false, tmproto.Header{})
-	if appVersion != v1 {
-		app.SetInitialAppVersionInConsensusParams(ctx, appVersion)
-		app.SetAppVersion(ctx, appVersion)
+func (app *App) OfferSnapshot(req *abci.RequestOfferSnapshot) (*abci.ResponseOfferSnapshot, error) {
+	app.Logger().Info("offering snapshot", "height", req.Snapshot.Height, "app_version", req.AppVersion)
+	if req.AppVersion != 0 {
+		if !isSupportedAppVersion(req.AppVersion) {
+			app.Logger().Info("rejecting snapshot because unsupported app version", "app_version", req.AppVersion)
+			return &abci.ResponseOfferSnapshot{
+				Result: abci.ResponseOfferSnapshot_REJECT,
+			}, nil
+		}
 	}
-	res.Timeouts.TimeoutCommit = app.getTimeoutCommit(appVersion)
-	res.Timeouts.TimeoutPropose = appconsts.GetTimeoutPropose(appVersion)
-	return res
```

Comment:
same question as above about adding the timeouts. My understanding is that we still need these but I could be wrong